crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+            let memory_index = index % self.max_memories;
+            let memory_start = self.memories_start + (index * self.memory_size);
+            let page_index = (addr - memory_start) / WASM_PAGE_SIZE;
+            let instance = self.get_instance(index / self.max_memories);
+
+            let init_page_index = instance
+                .memories
+                .get(
+                    DefinedMemoryIndex::from_u32(memory_index as u32)
+                        .try_into()
+                        .unwrap(),
+                )
+                .and_then(|m| {
+                    if page_index < m.size() as usize {
+                        Some(page_index)
+                    } else {
+                        None
+                    }
+                });
+
+            return Some(AddressLocation::MemoryPage {
+                page_addr: (memory_start + page_index * WASM_PAGE_SIZE) as _,
+                len: WASM_PAGE_SIZE,
+                instance,
+                memory_index,
+                page_index: init_page_index,
+            });
+        }
+
+        // Check for a table location
+        if addr >= self.tables_start && addr < self.tables_end {
+            let index = (addr - self.tables_start) / self.table_size;
+            let table_start = self.tables_start + (index * self.table_size);
+            let table_offset = addr - table_start;
+            let page_index = table_offset / self.page_size;
+
+            return Some(AddressLocation::TablePage {
+                page_addr: (table_start + (page_index * self.page_size)) as _,
+                len: self.page_size,
+            });
+        }
+
+        // Check for a stack location
+        if addr >= self.stacks_start && addr < self.stacks_end {
+            let index = (addr - self.stacks_start) / self.stack_size;
+            let stack_start = self.stacks_start + (index * self.stack_size);
+            let stack_offset = addr - stack_start;
+            let page_offset = (stack_offset / self.page_size) * self.page_size;
+
+            return Some(AddressLocation::StackPage {
+                page_addr: (stack_start + page_offset) as _,
+                len: self.page_size,
+                index,
+                guard_page: stack_offset < self.page_size,
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
+        .map_err(|e| format!("failed to change guard page protection: {}", e))?;
+
+    uffd.wake(page_addr as _, len).map_err(|e| {
+        format!(
+            "failed to wake page at {:p} with length {}: {}",
+            page_addr, len, e
+        )
+    })?;
+
+    Ok(())
+}
+
+unsafe fn initialize_wasm_page(
```

Comment:
While obvious in retrospect, could this be commented to indicate that it's only called with requests which are known to be valid wasm pages that must either have data copied in or data zeroed?

Reply:
I can add that comment.